xordiff.1
.\" Copyright (c) 2012 Renzo Davoli
.\"
.\" This is free documentation; you can redistribute it and/or
.\" modify it under the terms of the GNU General Public License,
.\" version 2, as published by the Free Software Foundation.
.\"
.\" The GNU General Public License's references to "object code"
.\" and "executables" are to be interpreted as the output of any
.\" document formatting or typesetting system, including
.\" intermediate and printed output.
.\"
.\" This manual is distributed in the hope that it will be useful,
.\" but WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\" GNU General Public License for more details.
.\"
.\" You should have received a copy of the GNU General Public
.\" License along with this manual; if not, write to the Free
.\" Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston,
.\" MA 02110-1301 USA.

.TH XORDIFF 1 "February 8, 2012" "Virtual Square Utilities"
.SH NAME
xordiff \- positional diff based on the exclusive or operation
.SH "SYNOPSIS"
.HP \w'\fBixordiff\fR\ 'u
\fBxordiff\fR [\fI-v\fR] [\fI-s\fR bufsize] [\fI-1234\fR] filea fileb file.a:b [file.ab:ba]

.SH "DESCRIPTION"
.PP
Each bit of the resulting file of
\fBxordiff\fR
command (file.a:b) contains the exclusive or (xor) of the corresponding
bits of filea and fileb. Given the following property of xor:
.nf
if (a xor b) is c then (a xor c) is b
.fi
the output file can be used as a diff file between the two input files.
\fBxordiff\fR itself can be used as the command to patch one of
the original files to get the other.
In fact for example after these two commands:
.in +4n
.nf
xordiff file1 file2 file1:2
xordiff file1 file1:2 newfile2
.fi
.in
file2 and newfile2 will have the same contents. \fBxordiff\fR has been designed
to guarantee the property above for any pair of files, including those having
different sizes.
The diff file has always the size of the second file. When the first file
is shorter the missing part of the file is considered zero, when the first
file is longer the part exceeding the size of the second file is ignored.
.br
Following the example above: 
.in +4n
.nf
xordiff file2 file1:2 file1onlyifsamesize
.fi
.in
file1onlyifsamesize will have the same contents of file1 only if file1 and
file2 have the same size.
.br
.sp
\fBxordiff\fR always generates sparse files when possible, so it is very
efficient (in terms of space) when the two files are very similar. This
is for example the case of different snapshots of a disk images used by 
virtual machines. Most of the data has not changed and its relative position
is unchanged.
.br
.sp
\fI-v\fR or \fI--verbose\fR provides a visible feedback of the 
process by typing one dot each 32MB processed and a line per GB.
.br
.sp
\fBxordiff\fR can compress/decompress data using \fBgzip(1)\fR or
\fBbzip2(1)\fR formats. To use this feature just add the proper suffix
to the filenames \fI.gz\fR or \fI.bz2\fR.
.br
.sp
One input file and/or one output file can be associated to the standard input
or to the standard output: use '-' as the filename. The following command 
updates a remote file.
.in +4n
.nf
xordiff file1 file2 - | ssh remhost xordiff remfile1 - remfile2
.fi
.in
It is possible to compress the stream, the following command updates
a remote copy of f1: f2 and remf2 will have the same contents:
.in +4n
.nf
xordiff -- f1 f2 -.bz2 | ssh remhost xordiff -- remf1 -.bz2 remf2
.fi
.in
the tag '--' is required to prevent getarg to parse -.bz2 as an argument.
.br
.sp
Usually \fBxordiff\fR is used to provide patches to upgrade (perhaps remote)
copies of large files subject to limited changes. As shown above \fBxordiff\fR
cannot be used for downgrade files if they have different sizes.
The forth optional file provide a (space optimized) general support for both 
upgrade and downgrade.
.br
The command:
.in +4n
.nf
xordiff file1 file2 file1:2 file12:21
.fi
.in
is functionally equivalent to:
.in +4n
.nf
xordiff file1 file2 file1:2
xordiff file2 file1 file2:1
xordiff file1:2 file2:1 file12:21
rm file2:1
.fi
.in
There is a significant saving in time in using one command instead of three,
expecially when the files are large, as all the files get sequentially read or
write just once. (The typical bottleneck of this command is the disk bandwidth
not the processor).

Like in the previous example file1:2 can be used to compute file2 given file1.
.in +4n
.nf
xordiff file1 file1:2 newfile2
.fi
.in
Now it is also possible to get file1 from file2 (downgrade) in the following
way:
.in +4n
.nf
xordiff file1:2 file12:21 - | xordiff file2 - file1
.fi
.in

If the size file2 is not smaller than the size of file1, file12:21 is an empty
file of the same size of file1. It uses no blocks on the file system 
being a sparse file, and can be compressed to a few bytes.
When file2 is smaller than file1, it has already the same size of file1,
it is empty in a first section large as the size of file2, then it contains
the final part of file1 (missing in file2). The same data is never stored twice
in  file1:2 and in  file12:21.
.SH SEE ALSO
sparsify(1), gzip(1), bzip2(1)
.SH AUTHORS
Howto's and further information can be found on the VirtualSquare Labs Wiki
<wiki.virtualsquare.org>.


xordiff.c
/*   
 *   Xordiff: Create a diff between two versions of the same file and store it
 *   as a sparse file
 *
 *   Copyright 2012 Renzo Davoli
 *   
 *   This program is free software; you can redistribute it and/or modify
 *   it under the terms of the GNU General Public License, version 3 or (at your option) 
 *   any later version, as published by the Free Software Foundation.
 *
 *   This program is distributed in the hope that it will be useful,
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *   GNU General Public License for more details.
 *
 *   You should have received a copy of the GNU General Public License along
 *   with this program; if not, write to the Free Software Foundation, Inc.,
 *   51 Franklin St, Fifth Floor, Boston, MA  02110-1301 USA.
 *
 */

#define _GNU_SOURCE 
#define _FILE_OFFSET_BITS 64
#define O_LARGEFILE
#include <stdio.h>
#include <fcntl.h>
#include <unistd.h>
#include <stdlib.h>
#include <string.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <getopt.h>
#include <mhash.h>
#include <bzlib.h>
#include <zlib.h>
#include <ioent.h>

#define STDBLOCKSIZE 4096
#define XOR_VERBOSE 0x1

unsigned inline long xordiff(unsigned long *b1, unsigned long *b2, unsigned long *b3,
		int bufsize)
{
	register int i;
	register unsigned long zero;
	for (i=zero=0; i<bufsize; i++) 
		zero |= (b3[i] = b1[i] ^ b2[i]);
	return zero;
}

unsigned inline long isnotzero(unsigned long *b, int bufsize)
{
	register int i;
	for (i=0; i<bufsize; i++)
		if (b[i])
			return 1;
	return 0;
}

void xorfile(struct ioent *f1, struct ioent *f2, struct ioent *fout, 
		struct ioent *fbiout, int blocksize, int flags)
{
	register off_t offset1;
	register off_t offset2;
	register int bufsize=blocksize / sizeof(unsigned long);
	register int verbose=flags & XOR_VERBOSE;
	unsigned long buf1[bufsize];
	unsigned long buf2[bufsize];
	unsigned long buf3[bufsize];
	unsigned long buf4[bufsize];
	int n1,n2;

	memset(buf4, 0, sizeof(long) * bufsize);
	for (offset1=offset2=0, n2=blocksize; n2 >= blocksize; offset1 += n1, offset2 += n2) {
		n1=f1->ft->ft_read(f1,buf1,blocksize);
		n2=f2->ft->ft_read(f2,buf2,blocksize);
		if (__builtin_expect(n1+n2 < 2*blocksize, 0)) {
			memset(((char *)buf1)+n1, 0, blocksize-n1);
			memset(((char *)buf2)+n2, 0, blocksize-n2);
		}
		//printf("off %lld %d %d %d\n",offset2,n1,n2,verbose);
		if (fbiout) {
			if (__builtin_expect(n1 > n2, 0)) {
				memcpy(((char *)buf4)+n2, ((char *)buf1)+n2, n1-n2);
				fbiout->ft->ft_write(fbiout,isnotzero(buf4,bufsize),buf4,n1,offset1);
			} else
				fbiout->ft->ft_write(fbiout,0,buf4,n1,offset1);
		} 
		fout->ft->ft_write(fout,xordiff(buf1,buf2,buf3,bufsize),buf3,n2,offset2);
		if (verbose) {
			if ((offset2 & 0xffffff)== 0) {
				if ((offset2 & 0x3fffffff)== 0x3f000000)
					fprintf(stderr, "X\n");
				else
					fprintf(stderr, ".");
			}
		}
	}
	fout->ft->ft_truncate(fout, offset2);
	//printf("%lld %lld\n",offset1,offset2);
	if (f1->hash || fbiout) {
		while ((n1=f1->ft->ft_read(f1,buf1,blocksize)) > 0) {
			if (fbiout) {
				if (__builtin_expect(n1 < blocksize,0)) 
					memset(((char *)buf1)+n1, 0, blocksize-n1);
				fbiout->ft->ft_write(fbiout,isnotzero(buf1,bufsize),buf1,n1,offset1);
				offset1 += n1;
				if (verbose) {
					if ((offset1 & 0xffffff)== 0) {
						if ((offset1 & 0x3fffffff)== 0x3f000000)
							fprintf(stderr, "X\n");
						else
							fprintf(stderr, ".");
					}
				}
			}
		}
		if (fbiout)
			fbiout->ft->ft_truncate(fbiout, offset1);
	}
	if (verbose)
		fprintf(stderr, "\n");
}

void usage(char *progname)
{
	fprintf(stderr,"Usage: %s [-v] [-s bufsize] {file1 | -} file2 filediff\n",progname);
	exit(1);

}

int main(int argc, char *argv[])
{
	static struct ioent f1, f2, fout, fbiout;
	static int flags;
	int blocksize=0;

	while (1) {
		int option_index = 0;
		int c;
		static struct option long_options[] = {
			{"help", no_argument, 0,  'h' },
			{"verbose", no_argument, 0,  'v' },
			{"bufsize", required_argument, 0,  's' },
			{0,         0,                 0,  0 }
		};

		c = getopt_long(argc, argv, "hvs:1234",
				long_options, &option_index);
		if (c == -1)
			break;

		switch(c) {
			case 's' : blocksize=atoi(optarg); break;
			case 'v': flags |= XOR_VERBOSE; break;
			case '1': f1.hash=mhash_init(MHASH_SHA1); break;
			case '2': f2.hash=mhash_init(MHASH_SHA1); break;
			case '3': fout.hash=mhash_init(MHASH_SHA1); break;
			case '4': fbiout.hash=mhash_init(MHASH_SHA1); break;
			case 'h': 
			default: usage(argv[0]);
		}
	}

	if (argc-optind < 3 || argc-optind > 4)
		usage(argv[0]);

	argc -= optind-1;
	argv += optind-1;

	open_ioent(&f1,argv[1],O_RDONLY,0);
	open_ioent(&f2,argv[2],O_RDONLY,0);
	open_ioent(&fout,argv[3],O_WRONLY|O_CREAT|O_EXCL,0666);
	if (blocksize == 0) {
		struct stat s;
		if (fout.ft != &ftfile || stat(argv[3],&s) < 0)
			blocksize = STDBLOCKSIZE;
		else
			blocksize = s.st_blksize;
	}
	if (argv[4]) {
		open_ioent(&fbiout,argv[4],O_WRONLY|O_CREAT|O_EXCL,0666);
		xorfile(&f1,&f2,&fout,&fbiout,blocksize,flags);
	} else
		xorfile(&f1,&f2,&fout,NULL,blocksize,flags);
	f1.ft->ft_close(&f1);
	f2.ft->ft_close(&f2);
	fout.ft->ft_close(&fout);
	if (argv[4])
		fbiout.ft->ft_close(&fbiout);
	if (f1.hash) printhash(f1.hash,"IN1",argv[1]);
	if (f2.hash) printhash(f2.hash,"IN2",argv[2]);
	if (fout.hash) printhash(fout.hash,"OUT",argv[3]);
	if (argv[4] && fbiout.hash) printhash(fbiout.hash,"ODX",argv[3]);
	return 0;
}
